mirrord/intproxy/src/proxies/incoming/interceptor.rs

```diff
@@ -0,0 +1,557 @@
+//! [`BackgroundTask`] used by [`Incoming`](super::IncomingProxy) to manage a single
+//! intercepted connection.
+
+use std::{
+    io::{self, ErrorKind},
+    net::SocketAddr,
+    time::Duration,
+};
+
+use hyper::{StatusCode, Version};
+use mirrord_protocol::tcp::{
+    HttpRequestFallback, HttpResponse, HttpResponseFallback, InternalHttpBody,
+};
+use thiserror::Error;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpSocket, TcpStream},
+    sync::oneshot::Receiver,
+    time,
+};
+
+use super::http::{HttpSender, UpgradedConnection};
+use crate::background_tasks::{BackgroundTask, MessageBus};
+
+/// Messages consumed by the [`Interceptor`] when it runs as a [`BackgroundTask`].
+pub enum MessageIn {
+    /// Request to be sent to the user application.
+    Http(HttpRequestFallback),
+    /// Data to be sent to the user application.
+    Raw(Vec<u8>),
+}
+
+/// Messages produced by the [`Interceptor`] when it runs as a [`BackgroundTask`].
+#[derive(Debug)]
+pub enum MessageOut {
+    /// Response received from the user application.
+    Http(HttpResponseFallback),
+    /// Data received from the user application.
+    Bytes(Vec<u8>),
```

Comment:
Let's either call both `Raw` or both `Bytes`.

I prefer `Raw` if it's `Raw(Vec<u8>)`, as `Bytes` might get confused with the `bytes` crate stuff.